checker/src/org/checkerframework/checker/linear/LinearAnnotatedTypeFactory.java

```diff
@@ -1,93 +1,88 @@
+        return getBundledTypeQualifiersWithPolyAll(Linear.class, Normal.class, Unusable.class);
     }
 
-    // TODO: Re-enable flow with the new org.checkerframework.dataflow framework.
+    @Override
+    public CFTransfer createFlowTransferFunction(
+            CFAbstractAnalysis<CFValue, CFStore, CFTransfer> analysis) {
+        return new LinearFlow(analysis);
+    }
 
-    /**
-     * Performs flow-sensitive analysis to mark reference types {@code Linear} as {@code Unusable}
-     * once they are used up.
-     *
-     * <p>A {code Linear} type is "used up" once the reference is mentioned, as an {@link
-     * IdentifierTree}.
-     */
-    /*
-    private class LinearFlow extends DefaultFlow<DefaultFlowState> {
-        public LinearFlow(BaseTypeChecker checker, CompilationUnitTree root,
-                Set<AnnotationMirror> annotations, AnnotatedTypeFactory factory) {
-            super(checker, root, annotations, factory);
-        }
+    private class LinearFlow extends CFTransfer {
+        private final AnnotatedTypeFactory factory;
 
-        /**
-         * Case 2: add {@code Unusable} to node type, if it is {@code Linear}.
-         *
-        @Override
-        public Void visitIdentifier(IdentifierTree node, Void p) {
-            super.visitIdentifier(node, p);
-            markAsUnusableIfLinear(node);
-            return null;
+        LinearFlow(CFAbstractAnalysis<CFValue, CFStore, CFTransfer> analysis) {
+            super(analysis);
+            factory = analysis.getTypeFactory();
         }
 
-        /**
-         * If the node is of type {@code Linear}, then transit its type
-         * into an {@code Unusable} type.
-         *
-         * The method should be called on every instance of a tree
-         * that causes the reference to be "used up".
-         *
-        private void markAsUnusableIfLinear(ExpressionTree node) {
-            if (!LinearVisitor.isLocalVarOrParam(node)) {
-                return;
-                }
+        @Override
+        public TransferResult<CFValue, CFStore> visitMethodInvocation(
+                MethodInvocationNode node, TransferInput<CFValue, CFStore> in) {
+            TransferResult<CFValue, CFStore> result = super.visitMethodInvocation(node, in);
 
-            Element elem = TreeUtils.elementFromUse(node);
-            assert elem != null;
-            if (this.flowState.vars.contains(elem)) {
-                int idx = this.flowState.vars.indexOf(elem);
-                if (this.flowState.annos.get(LINEAR, idx)) {
-                    this.flowState.annos.set(UNUSABLE, idx);
-                    this.flowState.annos.clear(LINEAR, idx);
+            @Nullable Tree receiverTree = node.getTarget().getReceiver().getTree();
+            if (receiverTree != null) {
+                AnnotatedTypeMirror type = factory.getAnnotatedType(receiverTree);
+                if (type.isAnnotatedInHierarchy(LINEAR)) {
```

Comment:
This will always return true.  (Well, technically if type is an AnnotatedTypeVariable it could return false.) . Did you mean `hasAnnotation`?

Reply:
This is what the old code was doing :) I tried to change as little as possible but looks like that strategy is a bad one.